ITE drivers/interrupt_controller: add wuc interface

RuibinChang opened the pull request:
Add wakeup controller interface for ITE it8xxx2 chip.

Signed-off-by: Ruibin Chang <ruibin.chang@ite.com.tw>

RuibinChang committed:
ITE drivers/interrupt_controller: add wuc interface

Add wakeup controller interface for ITE it8xxx2 chip.

Signed-off-by: Ruibin Chang <ruibin.chang@ite.com.tw>

RuibinChang committed:
ITE drivers/kscan: use wakeup interface for WUC pins

Use wakeup interface for WUC pins in it8xxx2 kscan driver.

Signed-off-by: Ruibin Chang <Ruibin.Chang@ite.com.tw>

keith-zephyr commented on soc/riscv/riscv-ite/common/soc_dt.h:
```diff
@@ -41,4 +41,123 @@
+	DT_PHA(IT8XXX2_DT_INST_WUCCTRL_0(inst, idx), wucs, mask)
+
+/**
+ * @brief For it8xxx2, get a node identifier for a phandle in a wucctrl
```
Can you provide an example for how the `wucctrl-n` properties are used?

RuibinChang replied:
In middle of this comment has two example for wucctrl-0 and wucctrl-1:  
IT8XXX2_DT_WUCCTRL_BY_IDX(DT_NODELABEL(n), 0, 1) // DT_NODELABEL(bar)
IT8XXX2_DT_WUCCTRL_BY_IDX(DT_NODELABEL(n), 1, 0) // DT_NODELABEL(baz)
So shall I add more example?

keith-zephyr replied:
This [draft CL](https://crrev.com/c/3359143) is a good example showing how to use the `wucctrl` property.  Do you plan to create a PR for that change?

RuibinChang replied:
Yes, I will create a GPIO PR with WUC changes.

keith-zephyr commented on dts/riscv/it8xxx2-wuc-map.dtsi:
```diff
@@ -0,0 +1,377 @@
+/ {
+	/* Mapping between wuc bits and source device */
+	it8xxx2-wuc-map {
+		compatible = "ite,it8xxx2-wuc-map";
```
Will this `wuc-map` replace the `gpio_irqs[]` [table](https://github.com/zephyrproject-rtos/zephyr/blob/f4a0ea7b43eee4d2ee735ab6beccc68c9d40a7d0/drivers/gpio/gpio_ite_it8xxx2.c#L119)?

RuibinChang replied:
I think we still need gpio_irqs[].gpio_pin member to find out which GPIO_INT is fired in gpio_ite_isr(). And we needn't the reset of gpio_irqs[] members. 
There is a draft CL about GPIO changes for WUC:
https://chromium-review.googlesource.com/c/chromiumos/third_party/zephyr/+/3359139/3

keith-zephyr replied:
You could initialize the gpio_irqs[] table using the `it8xxx2-wuc-map`.  All the WUC related mappings would be contained in one place.
```
	it8xxx2-wuc-map {
		wuc_wu22: wu22 {
			wucs = <&wuc2 2 BIT(2)>; /* WUI2 */
			gpios = <&gpioc 4 0>;  /* GPC4 */
		};
```

Alternately you could add another cell to `wuc-cells`:
```
		wuc_wu22: wu22 {
			wucs = <&wuc2 2 BIT(2) 4>; /* WUI2, GPC4 */
		};
```

RuibinChang replied:
Ack.
I find that the gpio pin can be found by comparing the irq and interrupt property, so we needn't the gpio_irqs[] table.
https://chromium-review.googlesource.com/c/chromiumos/third_party/zephyr/+/3359139/4

RuibinChang commented on soc/riscv/riscv-ite/common/soc_common.h:
```diff
@@ -48,6 +48,14 @@ extern void ite_intc_irq_priority_set(unsigned int irq,
 #endif /* CONFIG_ITE_IT8XXX2_INTC */
 
+#if CONFIG_ITE_IT8XXX2_WUC
+extern void it8xxx2_wuc_enable(uint8_t group, uint8_t mask);
```
Done

RuibinChang commented on include/dt-bindings/interrupt-controller/it8xxx2-wuc.h:
```diff
@@ -0,0 +1,25 @@
+#define ZEPHYR_INCLUDE_DT_BINDINGS_INTERRUPT_CONTROLLER_IT8XXX2_WUC_H_
+
+/** WUC reserved register of reg property */
+#define RESERVED		0
```
Done

keith-zephyr reviewed (commented):
I think it would be helpful if you added this [change](https://crrev.com/c/3359137) as a 2nd commit to this PR.

SebastianBoe commented on soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series:
```diff
@@ -13,6 +13,9 @@ config ITE_IT8XXX2_INTC
 
+config ITE_IT8XXX2_WUC
+	default y
+
```
Is this necessary?

Wonẗ it be default y due to dt_nodelabel_has_compat,wuc,it8xxx2-wuc being true ?

RuibinChang replied:
Yes, it's necessary, or it shows:  undefined reference wuc driver functions when build.

SebastianBoe replied:
This should be removed as it is DT that should be determining the default value.

RuibinChang commented:
> I think it would be helpful if you added this [change](https://crrev.com/c/3359137) as a 2nd commit to this PR.

But the GPIO functions (ex. gpio_ite_pin_interrupt_configure()) are called by UART and eSPI drivers, and both driver haven't been ready to create PR to call WUC interface. I think GPIO PR should be created after UART and eSPI commit, so far there is a kscan 2nd commit as example.

SebastianBoe commented on drivers/interrupt_controller/Kconfig.it8xxx2:
```diff
@@ -9,3 +9,11 @@ config ITE_IT8XXX2_INTC
+	help
+	 This option enables the wakeup controller interface for IT8XXX2
+	 family.
+	 This is required for GPIO, KSCAN, etc., interrupt support.
```
def_bool FOO

is the same as

bool
default FOO

so it surprises me that you get undefined reference issues when you don't do

default y

elsewhere.

In this build that causes undefined reference issues, are you expecting $(dt_nodelabel_has_compat,wuc,it8xxx2-wuc)
to be true?

Is $(dt_nodelabel_has_compat,wuc,it8xxx2-wuc) ever true?

perhaps $(dt_nodelabel_has_compat,wuc,it8xxx2-wuc) is accidentally always false, and then we fix it with Kconfig.

If $(dt_nodelabel_has_compat,wuc,it8xxx2-wuc) is always false, then there is no point in using DT.

SebastianBoe replied:
You could test if dt is accidentally always false

by doing

def_bool n

and see if anything changes. If it does, then I am wrong, and I apologize.

RuibinChang replied:
I upload my test code in https://github.com/zephyrproject-rtos/zephyr/commit/2028c92cc7eebba08a27f2180718796b26d04756 
to test "def_bool n", and it shows "undefined reference wuc driver functions", too, when build.  

No problem! Thank you for your reviewing on this PR.

SebastianBoe replied:
Sorry, I did not make myself clear.

Replacing $(dt_nodelabel_has_compat,wuc,it8xxx2-wuc) with n

and retaining the

config ITE_IT8XXX2_WUC
	default y

without there being any functional changes will prove that $(dt_nodelabel_has_compat,wuc,it8xxx2-wuc) is always false, which is not intended behaviour.

In your test code you also removed the config y, which does not prove anything interesting.

SebastianBoe replied:
Right, so this proves that $(dt_nodelabel_has_compat,wuc,it8xxx2-wuc) is evaluating to false.

Probably there is something wrong with the DT, because it should evaluate to true, right?

I have asked in Discord and found that

config ITE_IT8XXX2_WUC
default y

should be removed.

and that def_bool $(dt_nodelabel_has_compat,wuc,it8xxx2-wuc) should be used.

But you need to find out why the DT is being evaluated to false when it is expected to be true.

RuibinChang replied:
Done

SebastianBoe reviewed (changes requested):
DT code appears to be wrong

SebastianBoe reviewed (dismissed):
Kconfig is OK

keith-zephyr commented on drivers/interrupt_controller/wuc_ite_it8xxx2.c:
```diff
@@ -0,0 +1,161 @@
+void it8xxx2_wuc_set_polarity(uint8_t group, uint8_t mask, uint32_t flags)
+{
+	const struct it8xxx2_wuc_cfg *config =
+		(const struct it8xxx2_wuc_cfg *)wuc_dev[(group - 1)]->config;
```
If you change the first parameter to `const struct device *dev`, then you don't have to do the lookup from the `wuc_dev[]` array.

The wuc-cells only needs to be 1 to store the mask.

RuibinChang replied:
I tried and got compiler warning:
'struct device' declared inside parameter list will not be visible outside of this definition or declaration in soc_common.h.

The struct device declaration is after wuc function declaration, 
so in 3rd commit I include "device.h" in soc_common.h , then the compiler shows:
error: 'ENODEV' undeclared in device.h

and

lots warning: parameter names (without types) in function declaration in all of device_extern.h functions 
ex.
    8 | extern const struct device DEVICE_DT_NAME_GET(DT_N); /* dts_ord_0 */
      |                       ^~~~~~
   9 | extern const struct device DEVICE_DT_NAME_GET(DT_N_S_aliases); /* dts_ord_1 */
      |                      ^~~~~~

Shall I look into the warnings or just use the group cell ?

keith-zephyr replied:
Please look into the warnings.  It's a better implementation if you don't have to specify the group cell property.

RuibinChang replied:
Done

keith-zephyr commented on dts/bindings/interrupt-controller/ite,it8xxx2-wuc.yaml:
```diff
@@ -0,0 +1,28 @@
+    description: number of items to expect in an wakeup controller specifier
+
+wuc-cells:
+  - group
```
See above - the group # is redundant with the wuc phandle.

RuibinChang replied:
Done

keith-zephyr commented on drivers/kscan/kscan_ite_it8xxx2.c:
```diff
@@ -192,11 +200,16 @@ static bool read_keyboard_matrix(const struct device *dev, uint8_t *new_state)
+	 * NOTE: We want to clear the status as soon as possible,
+	 *       so clear KSI[7:0] pins at a time.
+	 */
+	it8xxx2_wuc_clear_status(config->wuc_map_list[0].group,
```
This call becomes `it8xxx2_wuc_clear_status(config->wuc_map_list[0].wucs, KEYBOARD_KSI_PIN_MASK);` after changing the `it8xxx2_wuc_clear_status` prototype.

RuibinChang replied:
Done

keith-zephyr reviewed (dismissed):
Thanks for fixing the wucctrl nodes.

RuibinChang commented:
> Thanks for fixing the wucctrl nodes.

Sure, group is really a redundant cell, and we save the lookup table times, thanks for your review.

RuibinChang commented:
Shall we solve the scan code error? Or could we merge the PR directly?  

There is a warning in scan the code step:
WARNING: You are using ScanCode Toolkit version 3.1.1, however the newer version 30.1.0 is available.
You should download and install the latest version of ScanCode with bug and security fixes and the latest license detection data for accurate scanning.
Visit https://github.com/nexB/scancode-toolkit/releases for details. Will check for missing copyrights...

Then in verify step:
scan/dts/riscv/it8xxx2-wuc-map.dtsi is not apache-2.0 licensed and has non-permissive license.

So it seems to need to update the ScanCode Toolkit .

After this PR is merged, there will be PRs of uart, eSPI and GPIO driver calling the wuc interface.

keith-zephyr commented on dts/riscv/it8xxx2-wuc-map.dtsi:
```diff
@@ -0,0 +1,377 @@
+/*
+ * Copyright (c) 2022 ITE Corporation. All Rights Reserved.
```
I think you need to delete the "All Rights Reserved" to fix the license check.

RuibinChang replied:
I had tried this before, I rebase the commits and try again, it still  isn't apache-2.0 licensed and has non-permissive license.

Actually other ITE files all have the "All Rights Reserved" string.

I confirm that the copyright format and short identifier is same as Standard License Header, defined in: 
https://spdx.org/licenses/Apache-2.0.html

mbolivar-nordic commented on dts/bindings/kscan/ite,it8xxx2-kscan.yaml:
```diff
@@ -14,6 +14,12 @@ properties:
     interrupts:
       required: true
 
+    wucctrl-0:
+      type: phandles
+      description: |
+        Pin configurations for the state. Content is specific to the
+        selected pin controller interface implementation.
```
This looks a bit odd. If it's pinctrl-related, why isn't this using the standard `pinctrl-0` property?

CC @gmarull

RuibinChang replied:
Wakeup controller interface is used to set that when the interrupt is triggered in EC low power mode, it can wakeup EC or not. Via this interface, we set the wakeup trigger edge, enable and disable for the specific pin which may be gpio pins or alternate pins. Wakeup controller interface is specific for ITE and the register block is different from pinmux and gpio. I update the description.

mbolivar-nordic commented on dts/bindings/interrupt-controller/ite,it8xxx2-wuc.yaml:
```diff
@@ -0,0 +1,27 @@
+  "wakeup-controller":
+    type: boolean
+    required: true
+    description: convey's this node is an wakeup controller
```
```suggestion
    description: conveys that this node is a wakeup controller
```

RuibinChang replied:
Done

mbolivar-nordic commented on dts/bindings/interrupt-controller/ite,it8xxx2-wuc.yaml:
```diff
@@ -0,0 +1,27 @@
+  "#wuc-cells":
+    type: int
+    required: true
+    description: number of items to expect in an wakeup controller specifier
```
```suggestion
    description: number of items to expect in an wakeup controller specifier
    const: 1
```

RuibinChang replied:
Done

mbolivar-nordic commented on soc/riscv/riscv-ite/common/soc_dt.h:
```diff
@@ -41,4 +41,120 @@
+ * @return node identifier for the phandle at index 'idx' in 'wucctrl-'pc_idx''
+ */
+#define IT8XXX2_DT_WUCCTRL_BY_IDX(node_id, pc_idx, idx)			\
+	DT_CAT6(node_id, _P_wucctrl_, pc_idx, _IDX_, idx, _PH)
```
Sorry, but this is an abstraction violation in multiple ways:

- The `DT_CAT<n>` macros are for internal use only within devicetree.h and its sub-headers; they are not public API. That is why they are not used anywhere outside of those headers right now. See `@internal` comment here: https://github.com/zephyrproject-rtos/zephyr/blob/ae8745df6ff5ad0b15b853a3ff271bfa2d26486e/include/devicetree.h#L3029.
- This is using the generated devicetree macros outside of devicetree.h and its sub-headers; the generated macros are private to those headers, and should not be used outside of them. See https://docs.zephyrproject.org/latest/guides/dts/api-usage.html#generated-macros, in particular "they should be considered an implementation detail"

We need to figure out another way to do this that preserves API boundaries.

This may take some time as we currently have special-cased `pinctrl-<N>` in gen_defines.py, so if you really need to do it this way with `wucctrl-<N>`, then we will need to figure out a way to generalize this special case.

Similar comments apply elsewhere.

RuibinChang replied:
Done

mbolivar-nordic commented on dts/bindings/kscan/ite,it8xxx2-kscan.yaml:
```diff
@@ -14,6 +14,12 @@ properties:
+    wucctrl-0:
+      type: phandles
+      description: |
+        configure wakeup controller that the interrupt of specific pin
+        can wakeup low power mode EC or not.
```
I'm sorry, but I'm having trouble understanding this updated description; could you please try to rephrase?

RuibinChang replied:
Done

mbolivar-nordic reviewed (changes requested):
Thanks for the response and the updates.

While I still don't understand the hardware enough to say what the right answer is, I took another look and I can say that this is not upstreamable in its current form. The main issue is around the abstraction violations in the devicetree API. See below comment for details.

Some other minor suggestions are included while I was reviewing.

gmarull commented on drivers/interrupt_controller/wuc_ite_it8xxx2.c:
```diff
@@ -0,0 +1,125 @@
+#include <logging/log.h>
+LOG_MODULE_REGISTER(wuc_ite_it8xxx2, LOG_LEVEL_ERR);
+
+#define DT_DRV_COMPAT ite_it8xxx2_wuc
```
by convention: please move this to the top of the file

RuibinChang replied:
Done

gmarull commented on drivers/interrupt_controller/wuc_ite_it8xxx2.c:
```diff
@@ -0,0 +1,125 @@
+#include <stdlib.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(wuc_ite_it8xxx2, LOG_LEVEL_ERR);
```
please register a new logging module in Kconfig, don't hardcode levels

RuibinChang replied:
Done, but I'm not sure the change satisfies your suggestion or not. 
Or could we do like this?
#define LOG_LEVEL       CONFIG_WUC_LOG_LEVEL
LOG_MODULE_REGISTER(wuc_ite_it8xxx2);

gmarull commented on drivers/interrupt_controller/wuc_ite_it8xxx2.c:
```diff
@@ -0,0 +1,125 @@
+void it8xxx2_wuc_enable(const struct device *dev, uint8_t mask)
+{
+	const struct it8xxx2_wuc_cfg *config = dev->config;
+	volatile uint8_t *reg_wuenr = (uint8_t *)config->reg_wuenr;
```
why not make `reg_wuenr` directly a pointer?

RuibinChang replied:
Done

gmarull commented on dts/riscv/it8xxx2-wuc-map.dtsi:
```diff
@@ -0,0 +1,377 @@
+/*
+ * Copyright 2022 ITE Corporation
```
why is this file at the root of `dts/riscv`?

RuibinChang replied:
Done, move this file to dts/riscv/ite/

gmarull commented on include/drivers/interrupt_controller/wuc_ite_it8xxx2.h:
```diff
@@ -0,0 +1,49 @@
+#ifndef ZEPHYR_DRIVERS_INTERRUPT_CONTROLLER_IT8XXX2_WUC_H_
+#define ZEPHYR_DRIVERS_INTERRUPT_CONTROLLER_IT8XXX2_WUC_H_
+
+#include <device.h>
```
`stdint.h` missing here

RuibinChang replied:
Done

gmarull commented on include/dt-bindings/interrupt-controller/it8xxx2-wuc.h:
```diff
@@ -0,0 +1,26 @@
+ * @{
+ */
+/** WUC rising edge trigger mode */
+#define WUC_TYPE_EDGE_RISING	(1U << 0)
+/** WUC falling edge trigger mode */
+#define WUC_TYPE_EDGE_FALLING	(1U << 1)
```
nit: `BIT(...)`

RuibinChang replied:
Done

gmarull commented on soc/riscv/riscv-ite/common/soc_dt.h:
```diff
@@ -41,4 +41,61 @@
+ *         property of that DT_DRV_COMPAT instance
+ */
+#define IT8XXX2_DT_INST_WUCCTRL_0(inst, idx)				\
+	DT_PHANDLE_BY_IDX(DT_DRV_INST(inst), wucctrl_0, idx)
```
`DT_INST_PHANDLE_BY_IDX`

RuibinChang replied:
Done

gmarull commented on dts/bindings/kscan/ite,it8xxx2-kscan.yaml:
```diff
@@ -14,6 +14,15 @@ properties:
     interrupts:
       required: true
 
+    wucctrl-0:
```
why is the `-0` index notation used? can there be `-N` entries?

RuibinChang replied:
Done, remove -0 index.

gmarull commented on drivers/interrupt_controller/Kconfig.it8xxx2:
```diff
@@ -9,3 +9,19 @@ config ITE_IT8XXX2_INTC
+	 family.
+	 This is required for KSCAN, UART, eSPI, GPIO etc., interrupt support.
+
+config WUC_LOG_LEVEL
+	int "WUC_LOG_LEVEL"
+	default 1
```
Add something like this to `drivers/interrupt_controller/Kconfig` and use it instead, 


https://github.com/zephyrproject-rtos/zephyr/blob/0c830822aee87b350025de35c3ba631ad834059d/drivers/can/Kconfig#L16-L18

RuibinChang replied:
Done

gmarull commented on drivers/interrupt_controller/Kconfig:
```diff
@@ -45,6 +45,14 @@ config LEON_IRQMP
 
+if ITE_IT8XXX2_WUC
+
+module = WUC
+module-str = wuc
```
better make it generic for all intc drivers, so maybe:

```suggestion
module = INTC
module-str = intc
```

RuibinChang replied:
Ok

gmarull commented on drivers/interrupt_controller/Kconfig.it8xxx2:
```diff
@@ -9,3 +9,15 @@ config ITE_IT8XXX2_INTC
+config ITE_IT8XXX2_WUC
+	bool "ITE it8xxx2 Wakeup controller (WUC) interface"
+	depends on SOC_IT8XXX2
+	default $(dt_nodelabel_has_compat,wuc1,$(DT_COMPAT_ITE_IT8XXX2_WUC))
```
why not:

```suggestion
	default $(dt_compat_enabled,$(DT_COMPAT_ITE_IT8XXX2_WUC))
```

?

RuibinChang replied:
Done

gmarull commented on drivers/interrupt_controller/wuc_ite_it8xxx2.c:
```diff
@@ -0,0 +1,126 @@
+#include <stdlib.h>
+
+#include <logging/log.h>
+#define LOG_LEVEL CONFIG_WUC_LOG_LEVEL
+LOG_MODULE_REGISTER(wuc_ite_it8xxx2);
```
```suggestion
LOG_MODULE_REGISTER(wuc_ite_it8xxx2, CONFIG_INTC_LOG_LEVEL);
```

RuibinChang replied:
Ok, thanks!

gmarull commented on include/dt-bindings/interrupt-controller/it8xxx2-wuc.h:
```diff
@@ -0,0 +1,26 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_INTERRUPT_CONTROLLER_IT8XXX2_WUC_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_INTERRUPT_CONTROLLER_IT8XXX2_WUC_H_
+
+/** WUC reserved register of reg property */
```
missing: `dt-bindings/dt-util.h` (for `BIT`)

RuibinChang replied:
Done

RuibinChang commented:
Rebase the commit

RuibinChang commented:
@mbolivar-nordic  Could you review the requested changes? Thanks~

keith-zephyr commented on dts/riscv/ite/it8xxx2-wuc-map.dtsi:
```diff
@@ -0,0 +1,377 @@
+			wucs = <&wuc15 BIT(0)>; /* GPL0 */
+		};
+		wuc_wu137: wu137 {
+			wucs = <&wuc15 BIT(1)>; /* GPL1 */
```
I think pin names `GPL1`,  `GPL2`, and `GPL3` are causing the license check failure.  I don't know if the scancode check can be bypassed or waived.

An alternative would be to document these pins as `GPIO L0`,`GPIO L1`, etc.

RuibinChang replied:
Done, thanks~

RuibinChang commented:
@mbolivar-nordic Could you review the changes that you requested? Thanks~

RuibinChang commented:
@mbolivar-nordic Could you review the changes that you requested, please?

mbolivar-nordic reviewed (approved):
Thanks for the devicetree updates!

RuibinChang commented:
Hi Caerles,

I rebased the commits, solved the kconfig conflict and changed  the deprecated UTIL_LISTIFY macro to LISTIFY, finally pushed with -f the PR successfully, but the PR is nothing changed on github. 


I also tried to solve the the kconfig conflict by the button "resolve conflicts" on github.  When I click "commit", it shows "This page is out of date", so the PR is still nothing changed.

Should I close the PR and re-open it?

RuibinChang commented:
Sorry, because this PR is nothing changed on github website when everytime I push the commit, I create a mirrored PR #43840.

carlescufi commented:
> Sorry, because this PR is nothing changed on github website when everytime I push the commit, I create a mirrored PR #43840.

Are you sure you are pushing to RuibinChang:wuc_interface ?

RuibinChang commented:
> > Sorry, because this PR is nothing changed on github website when everytime I push the commit, I create a mirrored PR #43840.
> 
> Are you sure you are pushing to RuibinChang:wuc_interface ?

Yes, I push to here 
https://github.com/zephyrproject-rtos/zephyr/compare/main...RuibinChang:wuc_interface?expand=1
It records those two commits pushed "on Mar 16, 2022".

We go to the first commit
https://github.com/zephyrproject-rtos/zephyr/commit/96d88081dddf2c35eb566dbb7170026ee439226c
It records the first commits with parent ID 2ef4750, the parent is the latest merged PR on Mar 16, 2022 morning. This commit solves kconfig conflict changes and uses LISTIFY macro instead. 

Then we go back to 
https://github.com/zephyrproject-rtos/zephyr/compare/main...RuibinChang:wuc_interface?expand=1
to "view pull request", there is nothing changed in "file changed".

carlescufi commented:
> > > Sorry, because this PR is nothing changed on github website when everytime I push the commit, I create a mirrored PR #43840.
> > 
> > 
> > Are you sure you are pushing to RuibinChang:wuc_interface ?
> 
> Yes, I push to here https://github.com/zephyrproject-rtos/zephyr/compare/main...RuibinChang:wuc_interface?expand=1 It records those two commits pushed "on Mar 16, 2022".
> 
> We go to the first commit [96d8808](https://github.com/zephyrproject-rtos/zephyr/commit/96d88081dddf2c35eb566dbb7170026ee439226c) It records the first commits with parent ID [2ef4750](https://github.com/zephyrproject-rtos/zephyr/commit/2ef47509c37163a0848b72b6ab37b7dbf831661e), the parent is the latest merged PR on Mar 16, 2022 morning. And we can observe that this commit solves kconfig conflict and uses LISTIFY macro instead.
> 
> Then we go back to https://github.com/zephyrproject-rtos/zephyr/compare/main...RuibinChang:wuc_interface?expand=1 to "view pull request", there is nothing changed in "file changed" (confilct alive and still use UTIL_LISTIFY).

OK, thank you for the explanation. I do not fully understand why this is happening, but let´s close this PR in benefit of #43840
